oadp-operator/controllers/registry.go

```diff
@@ -0,0 +1,353 @@
+			return nil
+		})
+
+		if err != nil {
+			return false, err
+		}
+
+		//TODO: Review registry deployment status and report errors and conditions
+
+		if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated {
+			// Trigger event to indicate registry deployment was created or updated
+			r.EventRecorder.Event(registryDeployment,
+				corev1.EventTypeNormal,
+				"RegistryDeploymentReconciled",
+				fmt.Sprintf("performed %s on registry deployment %s/%s", op, registryDeployment.Namespace, registryDeployment.Name),
+			)
+		}
+
+	}
+
+	return true, nil
+}
+
+// Construct and update the registry deployment for a bsl
+func (r *VeleroReconciler) buildRegistryDeployment(registryDeployment *appsv1.Deployment, bsl *velerov1.BackupStorageLocation) *appsv1.Deployment {
+	registryDeployment.Labels = r.getRegistryBSLLabels(bsl)
+
+	registryDeployment.Spec = appsv1.DeploymentSpec{
+		Replicas: pointer.Int32(1),
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					"component": constructBSLRegistryName(bsl),
+				},
+			},
+			Spec: corev1.PodSpec{
+				RestartPolicy: corev1.RestartPolicyAlways,
+				Containers:    r.buildRegistryContainer(bsl),
+			},
+		},
+	}
+	return registryDeployment
+}
+
+func (r *VeleroReconciler) getRegistryBSLLabels(bsl *velerov1.BackupStorageLocation) map[string]string {
+	labels := map[string]string{
+		"app.kubernetes.io/name":       OADPOperatorVelero,
+		"app.kubernetes.io/instance":   constructBSLRegistryName(bsl),
+		"app.kubernetes.io/managed-by": OADPOperator,
+		"app.kubernetes.io/component":  Registry,
+	}
+	return labels
+}
+
+func constructBSLRegistryName(bsl *velerov1.BackupStorageLocation) string {
+	return "oadp-" + bsl.Name + "-" + bsl.Spec.Provider + "-registry"
+}
+
+func (r *VeleroReconciler) buildRegistryContainer(bsl *velerov1.BackupStorageLocation) []corev1.Container {
+	containers := []corev1.Container{
+		{
+			Image: RegistryImage,
+			Name:  constructBSLRegistryName(bsl) + "-container",
+			Ports: []corev1.ContainerPort{
+				{
+					ContainerPort: 5000,
+					Protocol:      corev1.ProtocolTCP,
+				},
+			},
+			Env: getRegistryEnvVars(bsl),
+			LivenessProbe: &corev1.Probe{
+				Handler: corev1.Handler{
+					HTTPGet: &corev1.HTTPGetAction{
+						Path: "/v2/_catalog?n=5",
+						Port: intstr.IntOrString{IntVal: 5000},
+					},
+				},
+				PeriodSeconds:       5,
+				TimeoutSeconds:      3,
+				InitialDelaySeconds: 15,
+			},
+			ReadinessProbe: &corev1.Probe{
+				Handler: corev1.Handler{
+					HTTPGet: &corev1.HTTPGetAction{
+						Path: "/v2/_catalog?n=5",
+						Port: intstr.IntOrString{IntVal: 5000},
+					},
+				},
+				PeriodSeconds:       5,
+				TimeoutSeconds:      3,
+				InitialDelaySeconds: 15,
+			},
+		},
+	}
+
+	return containers
+}
+
+func getRegistryEnvVars(bsl *velerov1.BackupStorageLocation) []corev1.EnvVar {
+	envVar := []corev1.EnvVar{}
+	if bsl.Spec.Provider == AWSProvider {
```

Comment:
NIT: maybe make this a switch statement. Provider can't be all of these, so it should technically be `if; else` or switch